public/app/core/components/colorpicker/withColorPicker.tsx

```diff
@@ -0,0 +1,83 @@
+import React from 'react';
+import ReactDOM from 'react-dom';
+import Drop from 'tether-drop';
+import { SeriesColorPicker } from './SeriesColorPicker';
+
+export interface WithSeriesColorPickerProps {
+  color: string;
+  yaxis?: number;
+  optionalClass?: string;
+  onColorChange: (newColor: string) => void;
+  onToggleAxis?: () => void;
+}
+
+export default function withSeriesColorPicker(WrappedComponent) {
+  return class extends React.Component<WithSeriesColorPickerProps, any> {
+    pickerElem: any;
+    colorPickerDrop: any;
+
+    static defaultProps = {
+      optionalClass: '',
+      yaxis: undefined,
+      onToggleAxis: () => {},
+    };
+
+    constructor(props) {
+      super(props);
+      this.openColorPicker = this.openColorPicker.bind(this);
+    }
+
+    openColorPicker() {
+      if (this.colorPickerDrop) {
+        this.destroyDrop();
+      }
+
+      const { color, yaxis, onColorChange, onToggleAxis } = this.props;
+      const dropContent = (
+        <SeriesColorPicker color={color} yaxis={yaxis} onColorChange={onColorChange} onToggleAxis={onToggleAxis} />
+      );
+      const dropContentElem = document.createElement('div');
+      ReactDOM.render(dropContent, dropContentElem);
```

Comment:
Is this a suggested way of rendering drop element? Can't it be rendered declaratively?

Reply:
That's interesting. But I'm not sure how to use it here since Drop is doing a similar work. Could you suggest changes?

Reply:
This is interesting. But I'm not sure how to use it here since Drop is doing a similar work. Could you suggest changes?

Reply:
Sure thing @alexanderzobnin. Sorry for late answer, my comment was marked as outdated by GH.

First you would create something like dropdown wrapper component, that would serve as an entry point for the dropdown rendering. Best, if there was an element already available in DOM that would serve as a mount point for dropdowns. Having this, you would not have to worry about e.g. z-indexes as it's z axis positioning could be set upfront to be higher than the app's z-index in main css. In example below its `#dropdowns` element:

```tsx
type DropdownPortalProps = {
  children: JSX.Element;
}

function DropdownPortal(props: DropdownPortalProps) {
  return React.createPortal(
    this.props.children,
    document.getElementById("dropdowns")
  );
}
```

Having React abstraction over `Drop` is next step. 

It's responsibilities:
- creating actual instance of `Drop` when component is mounted
- destroying `Drop` instance when component is unmounted(btw you are missing this in current implementation)

This abstraction, let's name it `DropdownElement` would have an API that matches `Drop` config object. What it would render is just dropdown content wrapped in a `div` element that would serve as `Drop`'s `content` config property - you could retrieve it by React's `ref`

```tsx
// just a sketch of DropdownElement component
type DropdownElement = { 
  // match Drop config object type (without content)
  children: JSX.Element
}
class DropdownElement extends React.Component<DropdownElementProps> {
  constructor(props) {
    super(props);
    this.dropdownElement = React.createRef();
    this.dropInstance = null;
  }

  componentDidMount() {
    this.dropInstance = new Drop({
      ...this.props, // DropdownElement props would have the same shap as Drop config object
      content: this.dropdownElement.current
    });
  }

  componentWillUnmount() {
    if (this.dropInstance) {
      this.dropInstance.destroy();
    }
  }

  render() {
    return (
      <div ref={this.dropdownElement}>
        {this.props.children}
      </div>
    );
  }
}

```

Finally, one more abstraction to tie it all together. A `Dropdown` component, that would render `DropdownElement` within `DropdownPortal`:

```tsx
interface DropdownProps extends DropdownElementProps {} // open for future extension

function Dropdown({children, ...otherProps}: DropdownProps) {
  return (
    <DropdownPortal>
      <DropdownElement {...otherProps}>
        {children}
      </DropdownElement>
    </DropdownPortal>
  )
}
```
--- 

Having above infra you could remove dropdown creation logic from `withColorPicker` HOC. Instead of that you would only have a state that would be responsible for dropdown component rendering (simple on/off toggling with `onClick` event). `withColorPicker` render method would simply return: 

```tsx
<div
  className={optionalClass}
  ref={e => (this.pickerElem = e)}
  onClick={this.handleColorPickerToggle}
>
  <WrappedComponent {...wrappedComponentProps} />
  {this.state.isColorPickerOpen && (
    <Dropdown {pass dropdown config as props}>
      <SeriesColorPicker
        color={color}
        yaxis={yaxis}
        onColorChange={onColorChange}
        onToggleAxis={onToggleAxis}
      />
    </Dropdown>
  )}
</div>
```

This way the imperative nature of Drop is hidden and can be used declaratively in any context of Grafana's UI. Also, react's component lifecycle would be responsible for removing the `Drop` instances when these are no longer needed (component unmounted).

Yup, turned out to be a looooong comment;)